code/datums/attack_style/_blocking.dm

```diff
@@ -0,0 +1,212 @@
+	blocking_with = null
+	if(!QDELETED(linked_alert))
+		linked_alert.update_appearance(UPDATE_DESC)
+
+/datum/status_effect/blocking/proc/stop_blocking(obj/item/source)
+	SIGNAL_HANDLER
+
+	qdel(src)
+
+/datum/status_effect/blocking/proc/on_attacked(mob/living/source, atom/movable/hitby, damage, attack_text, attack_type, armour_penetration, damage_type)
+	SIGNAL_HANDLER
+
+	if(blocking_with)
+		if(!(blocking_with.can_block_flags & attack_type))
+			return NONE
+
+	else if(!(BLOCK_ALL_MELEE & attack_type))
+		return NONE
+
+	// Depending on the item (or lack thereof) you are blocking with, the damage taken is converted to more (or maybe less!) stamina damage
+	var/defense_multiplier = blocking_with ? blocking_with.get_blocking_ability(source, hitby, damage, attack_type, damage_type) : BARE_HAND_DEFENSE_MULTIPLIER
+	if(defense_multiplier < 0)
+		return NONE
+	if(damage_type == STAMINA)
+		// This is kinda a "anti-noobtrap" measure.
+		// If you are hit with an attack that does pure stamina damage (no side effects), like disabler fire,
+		// blocking it would serve you no gain and instead harm you by *increasing* the stamina damage you take.
+		// So instead of just disallowing users from blocking stamina attacks, we'll just cap it at 1x.
+		defense_multiplier = min(defense_multiplier, 1)
+
+	var/final_damage = defense_multiplier * damage
+	var/mob/living/attacker = GET_ASSAILANT(hitby)
+	if(istype(attacker) && HAS_TRAIT(attacker, TRAIT_HULK))
+		final_damage *= 1.25 // Hulk attacks are harder to stop
+	if(source.body_position == LYING_DOWN)
+		final_damage *= 1.25 // Harder to block while lying down
+	if(final_damage > 0)
+		// If you fail to take the damage, no block allowed
+		if(!source.apply_damage(final_damage, STAMINA, spread_damage = TRUE))
+			return NONE
+
+	// Stamcrit = failed
+	if(HAS_TRAIT(source, TRAIT_INCAPACITATED))
+		return NONE
+
+	// Stops all following effects of the attack.
+	if(isnull(blocking_with) || !blocking_with.on_successful_block(source, hitby, damage, attack_text, attack_type, damage_type))
+		source.visible_message(
+			span_danger("[source] blocks [attack_text][blocking_with ? " with [blocking_with]" : ""]!"),
+			span_danger("You block [attack_text][blocking_with ? " with [blocking_with]" : ""]!"),
+		)
+	if(final_damage > 0)
+		source.add_movespeed_modifier(/datum/movespeed_modifier/successful_block)
+		addtimer(CALLBACK(source, TYPE_PROC_REF(/mob, remove_movespeed_modifier), /datum/movespeed_modifier/successful_block), 0.5 SECONDS)
+	if(istype(attacker))
+		attacker.add_movespeed_modifier(/datum/movespeed_modifier/successful_block)
+		addtimer(CALLBACK(attacker, TYPE_PROC_REF(/mob, remove_movespeed_modifier), /datum/movespeed_modifier/successful_block), 0.8 SECONDS)
+		attacker.apply_damage(5, STAMINA, spread_damage = TRUE)
+
+	if(!QDELETED(src))
+		animate(shield_overlay, time = 0.15 SECONDS, pixel_x = 2, easing = BACK_EASING|EASE_OUT)
+		animate(time = 0.20 SECONDS, pixel_x = -2, easing = BACK_EASING|EASE_OUT)
```

Comment:
oh also it might be nice to make either the delay or the pixel offset here slightly random, +-1 kinda thing
could be wrong about that but a slight bit of noise helps with this sort of thing I think